cmd/checkpoint/main.go

```diff
@@ -324,59 +331,69 @@ func createCheckpointsForValidParents(client kubernetes.Interface, pods map[stri
 			continue
 		}
 
-		cp, err = checkpointSecretVolumes(client, cp)
-		if err != nil {
-			//TODO(aaron): This can end up spamming logs at times when api-server is unavailable. To reduce spam
-			//             we could only log error if api-server can't be contacted and existing secret doesn't exist.
-			glog.Errorf("Failed to checkpoint secrets for pod %s: %v", id, err)
-			continue
-		}
-		cp, err = checkpointConfigMapVolumes(client, cp)
+		cp, err = sanitizeCheckpointPod(cp)
 		if err != nil {
-			//TODO(aaron): This can end up spamming logs at times when api-server is unavailable. To reduce spam
-			//             we could only log error if api-server can't be contacted and existing configmap doesn't exist.
-			glog.Errorf("Failed to checkpoint configMaps for pod %s: %v", id, err)
+			glog.Errorf("Failed to sanitize manifest for %s: %v", id, err)
 			continue
 		}
 
-		cp, err = sanitizeCheckpointPod(cp)
+		podChanged, err := writeCheckpointManifest(cp)
 		if err != nil {
-			glog.Errorf("Failed to sanitize manifest for %s: %v", id, err)
+			glog.Errorf("Failed to write checkpoint for %s: %v", id, err)
 			continue
 		}
-		if err := writeCheckpointManifest(cp); err != nil {
-			glog.Errorf("Failed to write checkpoint for %s: %v", id, err)
+
+		// Check for secret and configmap changes if the pods have change or they haven't been checked in a while
+		if podChanged || lastCheckpoint.IsZero() || time.Since(lastCheckpoint) >= checkPointTimeout {
+
+			_, err = checkpointSecretVolumes(client, pod)
+			if err != nil {
+				//TODO(aaron): This can end up spamming logs at times when api-server is unavailable. To reduce spam
+				//             we could only log error if api-server can't be contacted and existing secret doesn't exist.
+				glog.Errorf("Failed to checkpoint secrets for pod %s: %v", id, err)
+				continue
+			}
+
+			lastCheckpoint = time.Now()
```

Comment:
This is a global timer but you're resetting it for the first pod you see, so the loop might not enter for the subsequent pods in this run. Perhaps create a struct to hold {pod, lastCheckpoint} pairs, or reset this after the loop runs?

Reply:
good catch